src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt

```diff
@@ -0,0 +1,157 @@
+    private val withTimetoken = 12345345452L
+    private val event: CapturingSlot<Event> = slot()
+
+    @BeforeEach
+    internal fun setUp() {
+        every { eventEngineManager.addEventToQueue(capture(event)) } returns Unit
+        objectUnderTest = Subscribe(eventEngineManager, subscriptionData)
+    }
+
+    @Test
+    internal fun `should store channels and channelGroups in local storage and pass SubscriptionChange event for handling when timeToken is zero`() {
+        val channelToSubscribe = CHANNEL_03
+        val channelGroupsToSubscribe = CHANNEL_GROUPS_03
+
+        objectUnderTest.subscribe(listOf(channelToSubscribe), listOf(channelGroupsToSubscribe), withPresence)
+
+        verify { eventEngineManager.addEventToQueue(any()) }
+
+        assertEquals(
+            Event.SubscriptionChanged(
+                (channelsInSubscriptionData + channelToSubscribe).toList(),
+                (channelGroupsInSubscriptionData + channelGroupsToSubscribe).toList()
+            ),
+            event.captured
+        )
+        assertTrue(subscriptionData.channels.contains(CHANNEL_01))
+        assertTrue(subscriptionData.channelGroups.contains(CHANNEL_GROUPS_01))
+    }
+
+    @Test
+    internal fun `should store channels and channelGroups in local storage and pass SubscriptionRestored event for handling when timeToken is non zero`() {
+        val channelToSubscribe = CHANNEL_03
+        val channelGroupsToSubscribe = CHANNEL_GROUPS_03
+
+        objectUnderTest.subscribe(
+            listOf(channelToSubscribe),
+            listOf(channelGroupsToSubscribe),
+            withPresence,
+            withTimetoken
+        )
+
+        verify { eventEngineManager.addEventToQueue(any()) }
+        assertEquals(
+            Event.SubscriptionRestored(
+                (channelsInSubscriptionData + channelToSubscribe).toList(),
+                (channelGroupsInSubscriptionData + channelGroupsToSubscribe).toList(),
+                SubscriptionCursor(withTimetoken, "42") // todo magic number
+            ),
+            event.captured
+        )
+        assertTrue(subscriptionData.channels.contains(CHANNEL_01))
+        assertTrue(subscriptionData.channelGroups.contains(CHANNEL_GROUPS_01))
+    }
+
+    @Test
+    internal fun `should remove channels and channelGroups from local storage and pass SubscriptionChange event for handling when at least one channel or channelGroup left in storage`() {
+        val channelsToUnsubscribe: List<String> = listOf(CHANNEL_01)
+        val channelGroupsToUnsubscribe: List<String> = listOf(CHANNEL_GROUPS_01)
+
+        objectUnderTest.unsubscribe(channelsToUnsubscribe, channelGroupsToUnsubscribe)
+
+        verify { eventEngineManager.addEventToQueue(any()) }
+        assertEquals(Event.SubscriptionChanged(listOf(CHANNEL_02), listOf(CHANNEL_GROUPS_02)), event.captured)
+        assertTrue(subscriptionData.channels.size == 1 && subscriptionData.channels.contains(CHANNEL_02))
```

Comment:
If you really have to verify internals there are better ways to write this assertion. 

```
assertEquals(setOf(CHANNEL_02), subscriptionData.channels)
```

Reply:
ok. I will modify it.